core/core.go

```diff
@@ -249,8 +254,14 @@ func (n *IpfsNode) teardown() error {
 	log.Debug("core is shutting down...")
 	// owned objects are closed in this teardown to ensure that they're closed
 	// regardless of which constructor was used to add them to the node.
-	closers := []io.Closer{
-		n.Repo,
+	var closers []io.Closer
+
+	// NOTE: the order that objects are added(closed) matters, if an object
+	// needs to use another during its shutdown/cleanup process, it should be
+	// closed before that other object
```

Comment:
ooof great point. maybe we should annotate each close line by what it's using like:

```
	// uses Exchange
	if n.FilesRoot != nil {
		closers = append(closers, n.FilesRoot)
 	}
```

or 

```
	// used by FilesRoot
	if n.Exchange != nil {
		closers = append(closers, n.Exchange)
 	}
```